src/anet.c

```diff
@@ -620,3 +620,43 @@ int anetFormatFdAddr(int fd, char *buf, size_t buf_len, int fd_to_str_type) {
     anetFdToString(fd,ip,sizeof(ip),&port,fd_to_str_type);
     return anetFormatAddr(buf, buf_len, ip, port);
 }
+
+/* Create a pipe buffer with given flags for read end and write end. */
+int anetPipe(int fds[2], int read_flags, int write_flags) {
+    int pipe_flags = 0;
+#if defined(__linux__) || defined(__FreeBSD__)
+    /* When it leverages the pipe2() to create pipe,
+     * there is no harm to set O_CLOEXEC to prevent fd leaks. */
+    pipe_flags = O_CLOEXEC | (read_flags & write_flags);
+    if (pipe2(fds, pipe_flags)) {
+        if (errno != ENOSYS)
+            return -1;
+        pipe_flags = 0;
+    } else if (read_flags == write_flags)
+        return 0;
+#endif
+
+    if (pipe_flags == 0 && pipe(fds))
+        return -1;
+
+    if (pipe_flags == 0 && (read_flags & O_CLOEXEC))
+        if (fcntl(fds[0], F_SETFD, O_CLOEXEC))
+            goto error;
+    if (pipe_flags == 0 && (write_flags & O_CLOEXEC))
+        if (fcntl(fds[1], F_SETFD, O_CLOEXEC))
+            goto error;
```

Comment:
This is slightly hard to understand. Just an idea. After successful pipe2() do this:

```C
    /* Clear the flags which we have already set using pipe2. */
    read_flags &= ~pipe_flags;
    write_flags &= ~pipe_flags;
```

Then the code setting flags is more simple:

```C
    if (read_flags & O_CLOEXEC)
        if (fcntl(fds[0], F_SETFD, O_CLOEXEC))
            goto error;
    if (write_flags & O_CLOEXEC)
        if (fcntl(fds[1], F_SETFD, O_CLOEXEC))
            goto error;
```

Reply:
Done.